pkg/translate/translateValue.go

```diff
@@ -0,0 +1,554 @@
+	// marshal value struct to yaml
+	valueYaml, err := yaml.Marshal(values)
+	if err != nil {
+		return nil, fmt.Errorf("error when marshalling value struct %v", err.Error())
+	}
+	// unmarshal yaml to untyped tree
+	var yamlTree = make(map[string]interface{})
+	err = yaml.Unmarshal(valueYaml, &yamlTree)
+	if err != nil {
+		return nil, fmt.Errorf("error when unmarshalling into untype tree %v", err.Error())
+	}
+
+	outputTree := make(map[string]interface{})
+	err = t.TranslateTree(yamlTree, outputTree, nil)
+	if err != nil {
+		return nil, err
+	}
+	outputVal, err := yaml.Marshal(outputTree)
+	if err != nil {
+		return nil, err
+	}
+
+	var cpSpec = &v1alpha2.IstioControlPlaneSpec{}
+	err = util.UnmarshalWithJSONPB(string(outputVal), cpSpec)
+
+	if err != nil {
+		return nil, fmt.Errorf("error when unmarshalling into control plane spec %v", err.Error())
+	}
+
+	return cpSpec, nil
+}
+
+// TranslateTree translates input value.yaml Tree to ControlPlaneSpec Tree
+func (t *ValueYAMLTranslator) TranslateTree(valueTree map[string]interface{}, cpSpecTree map[string]interface{}, path util.Path) error {
+	// translate enablement and namespace
+	err := t.setEnablementAndNamespacesFromValue(valueTree, cpSpecTree)
+	if err != nil {
+		return fmt.Errorf("error when translating enablement and namespace from value.yaml tree: %v", err.Error())
+	}
+	// translate with api mapping
+	err = t.translateTree(valueTree, cpSpecTree)
+	if err != nil {
+		return fmt.Errorf("error when translating value.yaml tree with global mapping: %v", err.Error())
+	}
+
+	// translate with k8s mapping
+	t.initK8SMapping(valueTree)
+	err = t.translateK8sTree(valueTree, cpSpecTree)
+	if err != nil {
+		return fmt.Errorf("error when translating value.yaml tree with kubernetes mapping: %v", err.Error())
+	}
+
+	// translate remaining untranslated paths into component values
+	err = t.translateRemainPaths(valueTree, cpSpecTree, nil)
+	if err != nil {
+		return fmt.Errorf("error when translating remaining path: %v", err.Error())
+	}
+	return nil
+}
+
+// setEnablementAndNamespaces translates the enablement and namespace value of each component in the baseYAML values
+// tree, based on feature/component inheritance relationship.
+func (t *ValueYAMLTranslator) setEnablementAndNamespacesFromValue(valueSpec map[string]interface{}, root map[string]interface{}) error {
+	for cnv, cni := range t.ValuesToComponentName {
+		enabled, err := name.IsComponentEnabledFromValue(cnv, valueSpec)
+		if err != nil {
+			return err
+		}
+		featureName := name.ComponentNameToFeatureName[cni]
+		tmpl := componentEnablementPattern
+		ceVal := renderFeatureComponentPathTemplate(tmpl, featureName, cni)
+		outCP := util.ToYAMLPath(ceVal)
+		// set component enablement
+		if err := tpath.WriteNode(root, outCP, enabled); err != nil {
+			return err
+		}
+		// set feature enablement
+		feVal := featureName + ".Enabled"
+		outFP := util.ToYAMLPath(string(feVal))
+		curEnabled, found, _ := name.GetFromTreePath(root, outFP)
```

Comment:
i have to fix this API - the fact of returning error for a missing path is really messing things up for the caller. could you please leave a TODO here to update the call when I make that change?

Reply:
but I am not sure whether I can use GetPathContext here, as my another comment said, GetPathContext doesn't work well with this case